sdk/messaging/azservicebus/internal/errors.go

```diff
@@ -229,17 +272,89 @@ func IsDrainingError(err error) bool {
 	return strings.Contains(err.Error(), "link is currently draining")
 }
 
-// IsSessionLockedError checks to see if this is the "you tried to get a session that was already locked"
-// error.
-func IsSessionLockedError(err error) bool {
-	var amqpError *amqp.Error
+var amqpConditionsToRecoveryKind = map[amqp.ErrorCondition]recoveryKind{
+	// no recovery needed, these are temporary errors.
+	amqp.ErrorCondition("com.microsoft:server-busy"):         RecoveryKindNone,
+	amqp.ErrorCondition("com.microsoft:timeout"):             RecoveryKindNone,
+	amqp.ErrorCondition("com.microsoft:operation-cancelled"): RecoveryKindNone,
+
+	// Link recovery needed
+	amqp.ErrorDetachForced: RecoveryKindLink, // "amqp:link:detach-forced"
+
+	// Connection recovery needed
+	amqp.ErrorConnectionForced: RecoveryKindConn, // "amqp:connection:forced"
+
+	// No recovery possible - this operation is non retriable.
+	amqp.ErrorMessageSizeExceeded:                                 RecoveryKindFatal,
+	amqp.ErrorUnauthorizedAccess:                                  RecoveryKindFatal, // creds are bad
+	amqp.ErrorNotFound:                                            RecoveryKindFatal,
+	amqp.ErrorNotAllowed:                                          RecoveryKindFatal,
+	amqp.ErrorInternalError:                                       RecoveryKindFatal, // "amqp:internal-error"
+	amqp.ErrorCondition("com.microsoft:entity-disabled"):          RecoveryKindFatal, // entity is disabled in the portal
+	amqp.ErrorCondition("com.microsoft:session-cannot-be-locked"): RecoveryKindFatal,
+	amqp.ErrorCondition("com.microsoft:message-lock-lost"):        RecoveryKindFatal,
+}
 
-	if !errors.As(err, &amqpError) {
-		return false
+func GetRecoveryKind(ctxForLogging context.Context, err error) recoveryKind {
+	if IsCancelError(err) {
+		return RecoveryKindFatal
+	}
+
+	var netErr net.Error
+
+	if errors.As(err, &netErr) {
```

Comment:
Remove

Reply:
The fall-through will basically mark this as RecoveryKindFatal, however. It seems like I should possibly classify it as RecoveryKindNone, indicating a retry would make sense.

Reply:
This goes back to that conversation with Jeff about retries.  The TL;DR is that unless you know for sure that an error is terminal, you should consider it as transient (this only works when there's a retry cap, which you have).

Reply:
Done.

Reply:
I still don't think there's a need to check for `net.Error`.  My pseudo-code algorithm in my head is "if err is not in this list of things we know to be terminal then retry".